### PR TITLE
GitHub Branch Protections Scaffolder Action

### PR DESCRIPTION
**Pull request by aaronnickovich**
#### :heavy_check_mark: Checklist

- [X] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [X] Added or updated documentation
- [X] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [X] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/scaffolder-backend-module-github-branch-protections/src/actions/githubBranchProtections.ts
Comment (vinzscam) [inline plugins/scaffolder-backend-module-github-branch-protections/src/actions/githubBranchProtections.ts:198]: if you do this, `handler` will always succeed without waiting until all the `client.graphql` calls have finished their work.
I guess you need to use a for each if your intention is to invoke `client.graphql` sequentially

**Review by vinzscam — CHANGES_REQUESTED** [review]

File: plugins/scaffolder-backend-module-github-branch-protections/src/actions/githubBranchProtections.ts
Comment (aaronnickovich) [reply plugins/scaffolder-backend-module-github-branch-protections/src/actions/githubBranchProtections.ts:198]: ah good point! Thanks @vinzscam, I will refactor

**Comment by benjdlambert** [comment]
Hey @aaronnickovich :wave:

First of all, thanks for this PR! :tada:

I've been thinking about this a little bit, and I think it probably makes more sense to actually call this package `@backstage/plugin-scaffolder-module-github` and have this as the new home for all GitHub related things, including moving over the existing github actions from the `scaffolder-backend` into this package too.

Do you think you could move this over?

**Comment by aaronnickovich** [comment]
> I've been thinking about this a little bit, and I think it probably makes more sense to actually call this package `@backstage/plugin-scaffolder-module-github` and have this as the new home for all GitHub related things, including moving over the existing github actions from the `scaffolder-backend` into this package too.
> 
> Do you think you could move this over?

Hey @benjdlambert, great idea! The scaffolder has an extensible module system. It makes sense to bring it out of the core backend code. This would be especially good for reducing security CVEs since not everyone will use every integration.

Also, I created a new RFC issue for tracking this larger effort. Hopefully you and other maintainers can provide any feedback or just approve the proposal if it looks OK: #21442

As mentioned in the RFC above, I think we should take a phased approach so we dont make breaking changes for each time github, gitlab, azure, gitea, etc. are migrated out of the core backend. For now, I think it makes sense to create a `@backstage/plugin-scaffolder-module-builtins` module that includes this newer `@backstage/plugin-scaffolder-module-github` module. The core backend scaffolder can just reference the builtins until we feel comfortable about breaking it out and letting adopters choose which packages to include.

**Comment by benjdlambert** [comment]
@aaronnickovich I went ahead and started to implement the RFC here: #21752

**Comment by aaronnickovich** [comment]
> @aaronnickovich I went ahead and started to implement the RFC here: #21752

Some priorities come up so thank you for your help @benjdlambert!

I still think it's useful to have additional branch protections so I'll update this PR soon.

**Comment by benjdlambert** [comment]
> I still think it's useful to have additional branch protections so I'll update this PR soon.

I agree with you, sounds like a good addition, and we were looking at building something like this internally @ Spotify too, so it's great that you're working on it! :tada:

> Also, I had rewritten most of the github module with Zod rather than json schemas which took up most of my time. Would you prefer I update the github module with Zod or do you prefer to keep the older json schemas?

I think it's up to you, you could also maybe move that out into a separate PR in order to not block the addition of this action to the github module? Makes it easier to review too?

Thanks for your help here! :tada:

File: plugins/scaffolder-backend-module-github-branch-protections/src/actions/githubBranchProtections.ts
Comment (vinzscam) [reply plugins/scaffolder-backend-module-github-branch-protections/src/actions/githubBranchProtections.ts]: mmm I don't think the issue is resolved as you are still using a `.forEach`

**Pull request closed without merging**
